File: C++/source/Framework/ObjectModel/WaveLocalObjectManagerForUserSpecificTasks.cpp
/***************************************************************************
 *   Copyright (C) 2005-2006 Vidyasagara Guntaka                           *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/ObjectModel/WaveLocalObjectManagerForUserSpecificTasks.h"
#include "Framework/Utils/WaveMutex.h"
#include <map>

namespace WaveNs
{

map<WaveServiceId, WaveServiceId> s_userSpecificServicesMap;
WaveMutex s_userSpecificServicesMapMutex;

WaveLocalObjectManagerForUserSpecificTasks::WaveLocalObjectManagerForUserSpecificTasks (const string &objectManagerName, const UI32 &stackSize, const vector<UI32> *pCpuAffinityVector)
    : WaveLocalObjectManager (objectManagerName, stackSize, pCpuAffinityVector)
{
    addToUserspecificServicesMap (getServiceId ());
}

WaveLocalObjectManagerForUserSpecificTasks::~WaveLocalObjectManagerForUserSpecificTasks ()
{
}

void WaveLocalObjectManagerForUserSpecificTasks::addToUserspecificServicesMap (const WaveServiceId &waveServiceId)
{
    s_userSpecificServicesMapMutex.lock ();
    s_userSpecificServicesMap[waveServiceId] = waveServiceId;
    s_userSpecificServicesMapMutex.unlock ();
}

bool WaveLocalObjectManagerForUserSpecificTasks::isAUserSpecificService (const WaveServiceId &waveServiceId)
{
    s_userSpecificServicesMapMutex.lock ();

    map<WaveServiceId, WaveServiceId>::iterator element = s_userSpecificServicesMap.find (waveServiceId);
    map<WaveServiceId, WaveServiceId>::iterator end     = s_userSpecificServicesMap.end ();
    bool                                          found   = false;

    if (element != end)
    {
        found = true;
    }
    else
    {
        found = false;
    }

    s_userSpecificServicesMapMutex.unlock ();

    return (found);
}

}

File: Java/source/com/CxWave/Wave/Framework/ObjectModel/WaveLocalObjectManagerForUserSpecificTasks.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.ObjectModel;

import java.util.HashMap;
import java.util.Map;

import com.CxWave.Wave.Framework.ObjectModel.Annotations.NonOM;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Framework.Utils.Synchronization.WaveMutex;

@NonOM
public class WaveLocalObjectManagerForUserSpecificTasks extends WaveLocalObjectManager
{
    private static final Map<WaveServiceId, WaveServiceId> s_userSpecificServicesMap      = new HashMap<WaveServiceId, WaveServiceId> ();
    private static final WaveMutex                         s_userSpecificServicesMapMutex = new WaveMutex ();

    public WaveLocalObjectManagerForUserSpecificTasks (final String waveLocalObjectManagerName)
    {
        super (waveLocalObjectManagerName);

        addToUserSpecificServicesMap (getServiceId ());
    }

    private static void addToUserSpecificServicesMap (final WaveServiceId waveServiceId)
    {
        s_userSpecificServicesMapMutex.lock ();

        s_userSpecificServicesMap.put (waveServiceId, waveServiceId);

        s_userSpecificServicesMapMutex.unlock ();
    }

    public static boolean isAUserSpecificService (final WaveServiceId waveServiceId)
    {
        s_userSpecificServicesMapMutex.lock ();

        final boolean isUserSpecificService = s_userSpecificServicesMap.containsKey (waveServiceId);

        s_userSpecificServicesMapMutex.unlock ();

        return (isUserSpecificService);
    }
}
